.github/workflows/build-test-distribute.yaml

```
@@ -170,3 +173,153 @@ jobs:
+          SINGLE_ENV='{"k8sVersion": ["${{ env.K8S_MAX_VERSION }}"], "target": ["multizone"], "arch":["amd64"]}'
+          ALL_DELTA_KDS='{"deltaKDS": [true]}'
+          ALL_CNI='{"cniNetworkPlugin": ["calico"]}'
+
+          # GitHub Actions does not support ARM64
+          EXCLUDE_ARM64='[{"arch": "arm64"}]'
+
+          # kind should only be used when testing ipv6 or with e2e-universal
+          EXCLUDE_KIND='[{"k8sVersion":"kind", "target": "kubernetes"}, {"k8sVersion":"kind", "target": "multizone"}]'
+
+          # universal only runs on kind
+          EXCLUDE_UNIVERSAL='[{"target":"universal", "k8sVersion":"${{ env.K8S_MIN_VERSION }}"}, {"target":"universal", "k8sVersion":"${{ env.K8S_MAX_VERSION }}"}]'
+
+          # non priority main e2e
+          EXCLUDE_MAIN_NON_PRIORITY='[{"k8sVersion":"kindIpv6"}, {"k8sVersion": "${{ env.K8S_MIN_VERSION }}"}]'
+
+          # Default matrix
+          SLOW=$(echo '{}' | $JQ_PATH -rc ".exclude += $EXCLUDE_ARM64" | $JQ_PATH -rc ". += $ALL_K8S_VERSION | . += $ALL_ARCH")
+          MAIN=$(echo '{}' | $JQ_PATH -rc ".exclude += $EXCLUDE_ARM64" | $JQ_PATH -rc ". += $ALL_K8S_VERSION | . += $ALL_ARCH | . += $ALL_TARGET" | $JQ_PATH -rc ".exclude += $EXCLUDE_KIND | .exclude += $EXCLUDE_UNIVERSAL" )
+          DELTA_KDS=$(echo '{}' | $JQ_PATH -rc ".exclude += $EXCLUDE_ARM64" | $JQ_PATH -rc ". += $ALL_DELTA_KDS | . += $SINGLE_ENV")
+          CNI=$(echo '{}' | $JQ_PATH -rc ".exclude += $EXCLUDE_ARM64" | $JQ_PATH -rc ". += $ALL_CNI | . += $SINGLE_ENV")
+
+          RUN_FULL_MATRIX="${{ contains(github.event.pull_request.labels.*.name, 'ci/run-full-matrix') }}"
+          if [[ "$RUN_FULL_MATRIX" == "false" ]]; then
+            echo "Skipping non priority e2e tests, because no label 'ci/run-full-matrix' was found on the pull request."
+
+            SLOW=''         # target==""
+            DELTA_KDS=''    # deltaKDS=="true"
+            CNI=''          # cniNetworkPlugin=="calico"
+            MAIN=$(echo $MAIN | $JQ_PATH -rc ".exclude += $EXCLUDE_ARM64 | .exclude += $EXCLUDE_MAIN_NON_PRIORITY")
+          fi
+
+          echo "slow=$SLOW" >> $GITHUB_OUTPUT
+          echo "main=$MAIN" >> $GITHUB_OUTPUT
+          echo "delta_kds=$DELTA_KDS" >> $GITHUB_OUTPUT
+          echo "cni=$CNI" >> $GITHUB_OUTPUT
+  e2e_slow:
+    name: "legacy-k8s:${{ matrix.arch }}-${{ matrix.k8sVersion }}"
+    runs-on: ubuntu-latest
+    needs: ["gen_e2e_matrix"]
+    if: needs.gen_e2e_matrix.outputs.slow != ''
+    strategy:
+      matrix: ${{ fromJSON(needs.gen_e2e_matrix.outputs.slow) }}
+    steps:
+      - run: |
+          echo "k8sVersion: ${{ matrix.k8sVersion }}"
+          echo "arch: ${{ matrix.arch }}"
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - uses: ./.github/actions/e2e
+        with:
+          k8sVersion: ${{ matrix.k8sVersion }}
+          arch: ${{ matrix.arch }}
+          parallelism: 3
+          target: ""
+  e2e_main:
```

```suggestion
  test/e2e_env:
```
This way it's the same as the code :)

---

Let's discuss on the call later and change accordingly in the second PR. 
Will that work?